draft-ietf-tls-esni.md

```diff
@@ -745,38 +745,24 @@ error code.
 
 ### Handling HelloRetryRequest {#client-hrr}
 
-As required in {{real-ech}}, clients offering ECH MUST ensure that all
-extensions or parameters that might change in response to receiving a
-HelloRetryRequest have the same values in ClientHelloInner and
-ClientHelloOuter. That is, if a HelloRetryRequest causes a parameter to be
-changed, the same change is applied to both ClientHelloInner and
-ClientHelloOuter. Applicable parameters include:
-
-1. TLS 1.3 {{!RFC8446}} ciphersuites in the ClientHello.cipher_suites list.
-1. The "key_share" and "supported_groups" extensions {{RFC8446}}. (These
-extensions may be copied from ClientHelloOuter into ClientHelloInner as
-described in {{real-ech}}.)
-1. Versions in the "supported_versions" extension, excluding TLS 1.2 and
-earlier. Note the ClientHelloOuter MAY include these older versions, while the
-ClientHelloInner MUST omit them.
-
-Future extensions that might change across first and second ClientHello messages
-in response to a HelloRetryRequest MUST have the same value.
-
-If the server sends a HelloRetryRequest in response to the ClientHello, the
-client sends a second updated ClientHello per the rules in {{RFC8446}}.
-However, at this point, the client does not know whether the server processed
-ClientHelloOuter or ClientHelloInner, and MUST regenerate both values to be
-acceptable. Note: if ClientHelloOuter and ClientHelloInner use different groups
-for their key shares or differ in some other way, then the HelloRetryRequest
-may actually be invalid for one or the other ClientHello, in which case a fresh
-ClientHello MUST be generated, ignoring the instructions in HelloRetryRequest.
-Otherwise, the usual rules for HelloRetryRequest processing apply.
-
-The client encodes the second ClientHelloInner as in {{encoding-inner}}, using
-the second ClientHelloOuter for any referenced extensions. It then encrypts
-the new EncodedClientHelloInner value as a second message with the previous
-HPKE context:
+When the server sends a HelloRetryRequest, the client determines if ECH was
+accepted by checking the message for an "encrypted_client_hello" extension with
+an 8-byte payload equal to `hrr_accept_confirmation` as defined in
+{{backend-server}}. If found, the client presumes acceptance and handles the
+HelloRetryRequest using ClientHelloInner. Otherwise, it presumes rejection and
+handles the HelloRetryRequest using ClientHelloOuter. Note that the
+client-facing server does not send "encrypted_client_hello" in case of
+rejection.
+
+#### Accepted ECH
+
+The client constructs the updated ClientHelloInner per the rules of {{RFC8446}}.
+It constructs the updated ClientHelloOuter as follows.
+
+The client encodes the updated ClientHelloInner as in {{encoding-inner}}, using
+the second ClientHelloOuter for any referenced extensions. It then encrypts the
```

Comment:
Or you could change both.  That isn't this PR, but it is straightforward.

So many words spilled on a feature that only really changes one field (key_share)...

Reply:
> Just to make sure I'm reading this right: the intent of this text is that you'd leave all other extensions of ClientHelloOuter unchanged, independent of what was in HelloRetryRequest?

The intention was that you would change the ClientHelloOuter as needed for the inner handshake. For example, if you need to change CHI.key_share, but you referenced this from CHO, you would update CHO.key_share.

>     * Don't compress the second ClientHelloInner at all. But this means wasting bytes because we're obligated to resend a bunch of stuff now known to be useless.

:-1: :)
 
>     * Allow the second ClientHelloOuter to arbitrarily change if the client feels like it. Then the client can arrange to compress the same way as it previously did (probably reusing the same code as above). But this means the rules on what the client does are weird and open-ended.

:+1: This is what I intended. Are you suggesting that the algorithm for updating CHI and CHO is ambiguous?

>     * Don't compress the second ClientHelloInner at all _and_ drop all extensions from the ClientHelloOuter because it's totally irrelevant at this point. This avoids wasting bytes and is well-defined, but it means yet another ad-hoc ClientHello serializer, albeit a pretty straightforward one.

I'd be OK with this.

>     * Relax RFC8446 4.1.2 rules. You compress the ones that didn't change and reorder as needed. Wastes less bytes but still some bytes. Also it means the client's compression logic gets even more interesting cases.

Which rules would we need to relax?